cpuid/src/transformer/mod.rs

```diff
@@ -64,7 +66,25 @@ impl VmSpec {
 pub enum Error {
     /// The maximum number of addressable logical CPUs cannot be stored in an `u8`.
     VcpuCountOverflow,
+    /// The max size has been exceeded
+    SizeLimitExceeded,
+    /// A call to an internal helper method failed
+    InternalError(super::common::Error),
 }
 
 pub type EntryTransformerFn =
     fn(entry: &mut kvm_cpuid_entry2, vm_spec: &VmSpec) -> Result<(), Error>;
+
+pub trait CpuidTransformer {
+    fn preprocess_cpuid(&self, _cpuid: &mut CpuId) -> Result<(), Error> {
+        Ok(())
+    }
+
+    fn transform_entry(
+        &self,
+        _entry: &mut kvm_cpuid_entry2,
+        _vm_spec: &VmSpec,
+    ) -> Result<(), Error> {
+        Ok(())
+    }
+}
```

Comment:
I find this two-step logic (pre-process entry list + transform individual entries) a bit difficult to follow.

Why not have a single transforming / filtering trait, that ingests a `CpuId` struct and outputs a "fixed" `CpuId`? Then, different vendors / templates / whatever else might affect cpuid, could provide an implementation for that trait, and we could chain these transformations together, to produce the final emulated cpuid.

I.e. the flow I'm imagining looks something like this:
```
vcpu::configure()
  ...
  vendor_filter <- pick_vendor_filter_based_on_host_cpu()
  template_filter <- pick_template_filter_based_on_microvm_config()
  emulated_cpuid <-
    template_filter.apply(
      vendor_filter.apply(
        get_emulated_cpuid_from_kvm()
      )
    )
  ...
```

Reply:
In general I agree with having a single transforming / filtering trait, that ingests a CpuId struct and outputs a "fixed" CpuId. It seems like a more consistent approach, but I'm not sure it would be that easy. There are more things to take into account here. Like the fact that certain templates can be applied only to certain vendors. For example C3 can only be applied to Intel.  I would like to dive deeper into it and take my time on implementing this. Also, since I have limited access to AMD hardware I would need this PR merged as soon as possible in order to continue the work on CPU feature masking for AMD.

So is it ok if I create another issue for this and address it probably next week ? We should also update the CPU templates in order to use the BitRange structure, I can do that as part of the same issue.

Reply:
Okay, then. We can take it offline and discuss this approach - I can't see many issues arising.

Reply:
I created #1025